feat(share_plus)!: Native share UI for Windows

alexmercerind opened the pull request:
## Description

This brings support for all the existing `package:share_plus` methods to Windows:
- `share`
- `shareFiles`
- `shareWithResult`
- `shareFilesWithResult`
- `shareXFiles`

This pull request adds the modern Share UI (_i.e._ `DataTransferManager`) for Windows 10 or greater.
It is very rich in experience & uses same UWP API through WRL.

The feature parity is equivalent as other platforms. Text, subject, images & files are supported.

On older Windows versions, existing behavior is kept (opening `mailto:` URL with subject & body).

## Demo

https://user-images.githubusercontent.com/28951144/193889028-b15f6c3c-9fc1-442c-b472-39b848fcf0ae.mp4

## Related Issues

This PR closes following issues:

- Closes #333. 
- Closes #999.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I **haven't** updated the version in `pubspec.yaml` and `CHANGELOG.md`.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

alexmercerind commented:
This is not exactly _a breaking change_.
But, since the change causes a big behavioural change from existing, I think it's best to ship it in a major version increment.

alexmercerind commented:
Adding integration tests & `withResult` implementation.

CodeDoctorDE commented:
The latest release uses XFiles, does it support it?

alexmercerind commented:
> The latest release uses XFiles, does it support it?

Web specific.

CodeDoctorDE commented:
Yeah, because of that. But shareFiles is now unsupported

alexmercerind commented:
> Yeah, because of that. But shareFiles is now unsupported

It still uses `shareFilesWithResult` internally. No changes have been introduced to native code of Android, iOS, macOS or Linux.

This Windows implementation is equivalent & still working similarly.

https://github.com/fluttercommunity/plus_plugins/blob/69fbff0248fa2e6f6baf20fd57575ad047f9aea4/packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart#L135-L152

miquelbeltran commented:
Thanks for taking care of the changes in main @alexmercerind !

alexmercerind commented:
If version `4.5.0` is decided to be merged into `5.0.0`, that's something we can do in `CHANGELOG`.

https://github.com/fluttercommunity/plus_plugins/blob/0d5069db5c4f508f3c995085b101337cc51d7238/packages/share_plus/share_plus/CHANGELOG.md?plain=1#L1-L11

miquelbeltran reviewed (commented):
Tentatively looks good :) But I am not a Windows dev so I cannot comment, any devs familiar with Windows want to give it a look and comment?

miquelbeltran commented:
Hey! Due to changes in our contributor guidelines, all PRs should now:

- Have a title that follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/) format. e.g. `feat(package_name)!: feature description`
- **DO NOT** modify the CHANGELOG.md or the version in the `pubspec.yaml`. This is going to be an automated process now.

We would ask you kindly to update the PR following these changes.

Thanks!

alexmercerind commented:
Hi!
I've removed any version changes in `pubspec.yaml` & edits to `CHANGELOG.md`.
Also, I've rebased & cleaned-up commit history to keep things tidy.

miquelbeltran commented:
Later today I will check this on my Windows machine

miquelbeltran commented:
I checked this locally and works flawlessly, let's go!

miquelbeltran commented on packages/share_plus/share_plus_windows/test/share_plus_windows_test.dart:
```diff
@@ -1,40 +1,72 @@
-  test('url encoding is correct for &', () async {
-    final mock = MockUrlLauncherPlatform();
-    UrlLauncherPlatform.instance = mock;
+  test(
```
Thanks a lot for including unit tests as well!

miquelbeltran commented:
Thanks for the work @alexmercerind !
